### PR TITLE
Shell - improvements in shell_execute_cmd

### PR DESCRIPTION
**Pull request by ceolin**

### CODE REVIEW

**Comment by jarz-nordic** [comment]
@ceolin : what is the reason to remove const?

**Comment by ceolin** [comment]
> @ceolin : what is the reason to remove const?

const mean that the function will not change that variable. This is not the case of this function, since it's copying the command to a buffer.

File: subsys/shell/shell.c [hunk 1]
```diff
@@ -1668,15 +1668,25 @@ static void help_options_print(const struct shell *shell,
 
 	/* Formating and printing all available options (except -h, --help). */
 	for (size_t i = 0; i < opt_cnt; ++i) {
+		size_t len = 0;
+		size_t buf_len = sizeof(shell->ctx->temp_buff);
 		if (opt[i].optname_short) {
-			strcpy(shell->ctx->temp_buff, opt[i].optname_short);
+			len += shell_strlen(opt[i].optname_short);
```
Comment (jarz-nordic) [inline subsys/shell/shell.c]: This is very good catch!

I am not sure if this code solves the problem. In my opinion if command buffer is not large enough to store option or command user will not have chance to use it. So it makes limited sense to print only partial option string with no indication it is not compleate.

I think we can implement either an ASSERT or shell error message that command buffer is too small. What do you think?

File: subsys/shell/shell.c [hunk 2]
```diff
@@ -1708,7 +1718,7 @@ static void help_subcmd_print(const struct shell *shell)
 			break;
 		}
 
-		u16_t len = shell_strlen(entry->syntax);
+		size_t len = shell_strlen(entry->syntax);
```
Comment (jarz-nordic) [inline subsys/shell/shell.c]: For `shell_strlen` related changes please use casting to u16_t.

**Review by jakub-uC — CHANGES_REQUESTED** [review]
Lets discuss about findings but `const` must stay ;)

File: subsys/shell/shell.c (hunk 2 above)
Comment (ceolin) [reply subsys/shell/shell.c]: If we are going to always cast the return of shell_strlen() why don't simply change this function to return u16_t ?

File: subsys/shell/shell.c [hunk 3]
```diff
@@ -1789,16 +1799,16 @@ int shell_cmd_precheck(const struct shell *shell,
...
 
-int shell_execute_cmd(const struct shell *shell, const char *cmd)
+int shell_execute_cmd(struct shell *shell, const char *cmd)
 {
-	u16_t cmd_len = shell_strlen(cmd);
+	size_t cmd_len = shell_strlen(cmd);
 
 	if (cmd == NULL) {
-		return -ENOEXEC;
+		return -EINVAL;
```
[the remark this answers is not in the document]
Comment (ceolin) [reply subsys/shell/shell.c]: Yeah, this is controversial, I thought this was the reason. But my 5 cents is that using a POSIX error with a different meaning is worse than use a more generic error. But we can left this as it is

File: subsys/shell/shell.c [hunk 4]
```diff
@@ -654,7 +654,7 @@ static void partial_autocomplete(const struct shell *shell,
 {
 	const char *completion;
 	u16_t common = common_beginning_find(cmd, &completion, first, cnt);
-	int arg_len = shell_strlen(arg);
+	size_t arg_len = shell_strlen(arg);
```
[the remark this answers is not in the document]
Comment (ceolin) [reply subsys/shell/shell.c]: commented in the end about this

File: include/shell/shell.h [hunk 5]
```diff
@@ -716,7 +716,7 @@ void shell_print_stream(const void *user_ctx, const char *data,
  *
  * @returns Result of the execution
  */
-int shell_execute_cmd(const struct shell *shell, const char *cmd);
+int shell_execute_cmd(struct shell *shell, const char *cmd);
```
[the remark this answers is not in the document]
Comment (ceolin) [reply include/shell/shell.h]: don't complete agree, but it is far :)

File: subsys/shell/shell.c (hunk 1 above)
Comment (ceolin) [reply subsys/shell/shell.c]: yeah, I was not sure about the behaviour here as well, As it is just to print a help, I thought would be better some information than nothing, but we can assert or return an error. What do you prefer ?

File: subsys/shell/shell.c (hunk 2 above)
Comment (jarz-nordic) [reply subsys/shell/shell.c]: this is very good idea!

File: subsys/shell/shell.c (hunk 1 above)
Comment (jarz-nordic) [reply subsys/shell/shell.c]: Below is my proposal, if buffer is too small shell will print warning and as many data as possible:

```
static void help_options_print(const struct shell *shell,
			       const struct shell_getopt_option *opt,
			       size_t opt_cnt)
{
	u32_t buf_len = sizeof(shell->ctx->temp_buff);
	static const char help_opt[] = "-h, --help";
	u32_t longest_name = shell_strlen(help_opt);
	static const char opt_sep[] = ", ";


	shell_fprintf(shell, SHELL_NORMAL, "Options:\r\n");

	if ((opt == NULL) || (opt_cnt == 0)) {
		help_item_print(shell, help_opt, longest_name,
				"Show command help.");
		return;
	}

	/* Looking for the longest option string. */
	for (size_t i = 0; i < opt_cnt; ++i) {
		u32_t len = shell_strlen(opt[i].optname_short) +
						 shell_strlen(opt[i].optname) +
						 shell_strlen(opt_sep);

		longest_name = len > longest_name ? len : longest_name;
	}

	if (longest_name > buf_len) {
		shell_fprintf(shell, SHELL_WARNING,
			      "Some options will not be printed correctly."
			      "Increase command buffer size.\r\n");
	}

	/* help option is printed first */
	help_item_print(shell, help_opt, longest_name, "Show command help.");

	/* prepare a buffer to compose a string:
	 *	option name short - option separator - option name
	 */
	memset(shell->ctx->temp_buff, 0, longest_name + 1);

	/* Formating and printing all available options (except -h, --help). */
	for (size_t i = 0; i < opt_cnt; ++i) {
		u32_t len = 0;
		if (opt[i].optname_short) {
			len += shell_strlen(opt[i].optname_short);
			strncpy(shell->ctx->temp_buff, opt[i].optname_short,
				min(len, buf_len));
		}
		if (opt[i].optname && (len < buf_len)) {
			if (*shell->ctx->temp_buff) {
				len += shell_strlen(opt_sep) +
					shell_strlen(opt[i].optname);
				if (len < buf_len) {
					strcat(shell->ctx->temp_buff, opt_sep);
					strcat(shell->ctx->temp_buff,
					       opt[i].optname);
				}
			} else {
				strncpy(shell->ctx->temp_buff, opt[i].optname,
					buf_len);
			}
		}
		help_item_print(shell, shell->ctx->temp_buff, longest_name,
				opt[i].optname_help);
	}
}
```

**Comment by ceolin** [comment]
@jarz-nordic patch updated, can you take a look once again ?

**Comment by jarz-nordic** [comment]
@ceolin May you please wait with the change until PR #11664 will be in?
I expect some merge conflicts. It will be easie to fix this PR due to scope of change.

**Comment by ceolin** [comment]
> @ceolin May you please wait with the change until PR #11664 will be in?
> I expect some merge conflicts. It will be easier to fix this PR due to scope of change.

sure

**Comment by ceolin** [comment]
> > @ceolin May you please wait with the change until PR #11664 will be in?
> > I expect some merge conflicts. It will be easier to fix this PR due to scope of change.
> 
> sure

@jarz-nordic can you approve this commit if there is nothing to change ?

**Review by jakub-uC — APPROVED** [review]

**Comment by jarz-nordic** [comment]
@ceolin : I've set DNM label because there are some merge conflicts and there will be more shortly.

btw. PR #11664 will remove function `help_options_print` completely. Options will be removed from help concept.

**Comment by nashif** [comment]
@ceolin can you please rebase?

**Comment by ceolin** [comment]
done :)

**Comment by ceolin** [comment]
@nashif  ping

**Comment by nashif** [comment]
> @nashif ping

pong

**Comment by ceolin** [comment]
@nashif  sorry, there was a problem in the commit msg. I've fixed it now.

File: subsys/shell/shell_utils.c [hunk 6]
```diff
@@ -197,8 +197,7 @@ char shell_make_argv(size_t *argc, char **argv, char *cmd, u8_t max_argc)
 void shell_pattern_remove(char *buff, u16_t *buff_len, const char *pattern)
 {
 	char *pattern_addr = strstr(buff, pattern);
-	u16_t pattern_len = shell_strlen(pattern);
-	size_t shift;
+	u16_t shift, pattern_len = shell_strlen(pattern);
```
[the remark this answers is not in the document]
Comment (ceolin) [reply subsys/shell/shell_utils.c]: sure !

**Pull request merged**